ufl/core/external_operator.py

```diff
@@ -0,0 +1,254 @@
+# -*- coding: utf-8 -*-
+"""This module defines the ``ExternalOperator`` class, which symbolically represents operators that are not straightforwardly expressible in UFL. A practical implementation is required at a later stage to define how this operator should be evaluated as well as its derivatives from a given set of operands.
+"""
+
+# Copyright (C) 2019 Nacime Bouziani
+#
+# This file is part of UFL (https://www.fenicsproject.org)
+#
+# SPDX-License-Identifier:    LGPL-3.0-or-later
+#
+# Modified by Nacime Bouziani, 2020
+
+from ufl.coefficient import Coefficient
+from ufl.core.operator import Operator
+from ufl.core.ufl_type import ufl_type
+from ufl.constantvalue import as_ufl
+from ufl.log import error
+from ufl.finiteelement.finiteelement import FiniteElement
+from ufl.finiteelement.mixedelement import VectorElement, TensorElement
+from ufl.functionspace import FunctionSpace
+from ufl.domain import default_domain
+
+
+@ufl_type(num_ops="varying", inherit_indices_from_operand=0, is_differential=True)
+class ExternalOperator(Operator):
+
+    # Slots are disabled here because they cause trouble in PyDOLFIN
+    # multiple inheritance pattern:
+    _ufl_noslots_ = True
+
+    _external_operator_type = ''
```

Comment:
What is this thing? It seems like it is used to specify certain behaviour of the operator. But it's just a random string that sometimes has a magic meaning.

Two options: 

1. make `ExternalOperator` a base class and the types inherit, e.g. `GlobalExternalOperator` etc..., then the handlers can deal with them

2. Use an enum and make sure the handlers explicitly handle all the cases 

```python
from enum import IntEnum

OperatorTypes(IntEnum):
   GLOBAL = 0
   LOCAL = 1

class ExternalOperator(Operator):
    Type = OperatorTypes

    operator_type = None

...

class ...(MultiFunction):
    def external_operator(...):
        typ = o.operator_type
        if not isinstance(typ, Enum) or typ not in o.Type:
           raise ValueError("Unknown operator type")
        if typ == o.Type.GLOBAL:
           ...
```

I think I prefer the former.